### PR TITLE
major refactor including new advertise functionality

### LINKED ISSUE #40: DICE has incorrect julian day calculation
The calculation for DICE is incorrect for the julian day - but this only appears after the first day boundary.
```
 call shr_cal_ymd2julian(0, current_mon, current_day, current_tod, jDay , calendar)    ! julian day for model
 call shr_cal_ymd2julian(0, 9,           1,           0,           jDay0, calendar)    ! julian day for Sept 1
 cosArg = 2.0_R8*pi*(jday - jday0)/365.0_R8
```
The problem is that current_mon current_day was being set for the previous time not the updated time (where all of the other data model calculations are taking place).

### PR DESCRIPTION
**Pull request by mvertens**
major refactor including new advertise functionality

This PR has several pieces:
NOTE: all prognostic components have been updated as well to incorporate the below functionality

1) NUOPC and MCT directories are now totally separate for the caps and the associated methods

2) There is no longer references to seq_timemgr_mod in the NUOPC cap directories

3) Restart alarms are now being set in the routine ModelSetRunClock
```
call NUOPC_CompAttributeGet(gcomp, name="restart_option", value=restart_option, rc=rc)
call NUOPC_CompAttributeGet(gcomp, name="restart_n"     , value=cvalue, rc=rc)
call NUOPC_CompAttributeGet(gcomp, name="restart_ymd"   , value=cvalue, rc=rc)
```
 - This has been implemented for all cime components and mom6
 - **NEXT STEP:** Once this is implemented in cam and clm, then the next step will be
remove the creation of all the clocks in esm.F90 and also remove the slot calls

4) New advertise methods have been implemented in ALL prognostic, data model and dead components
 - There is no longer calls to shr_nuopc_fldList_Realize, shr_nuopc_fldList_Concat
shr_nuopc_fldList_Deactivate, shr_nuopc_fldList_Getnumflds, shr_nuopc_fldList_Getfldinfo
 - The mediator still has these calls
 - **NEXT STEP**: create a new module that has the dictionary of all possible fields created
on all processors and the call to esmFlds_Init will only be done by the mediator

5) The nuopc data model implementations have been significantly
refactored to no longer hardwire the contents of the character
arrays avifld and avofld and new nuopc share code has been implemented to have more code reuse
 - **NEXT STEP**: continue moving duplicated code into share directory and aim to generalize code to not be cesm specific

6) Code base 
  - cime (3de9261)
  - cam (branch_tags/nuopc_cap_tags/nuopc_cap_n20_cam6_0_002)
  - cice (7fb6ac5) 
  - clm ( 88fc479)
  - MOM6 (64665e0)

TODO: NEW baselines need to be created

Test suite: 
Test baseline: restart_cime5.6.8
Test namelist changes: no
Test status: - except for tests with dice
**baseline failures:** 
```
    FAIL ERS_Vnuopc_Ld5.T62_g16.CMOM.cheyenne_intel.mom-nuopc_cap     
    FAIL SMS_Vnuopc_Ld5.T62_g16.CMOM.cheyenne_intel.mom-nuopc_cap       
    FAIL SMS_Vnuopc_Ld1.T31_g37_rx1.A.cheyenne_intel   
          **expected due to changes in dice for #40**                
    FAIL ERS_Vnuopc_Ld5.T62_g37.DTEST.cheyenne_intel.cice-nuopc_cap    
          ** expected since comparing against 5.6.8 where restart test was failing    
    FAIL ERS_Vnuopc_Ln9.f19_g16.X.cheyenne_intel                      
    FAIL SMS_Vnuopc.f19_g16.X.cheyenne_intel                         
         **expected due to order of operation differences**
    FAIL ERS_Vnuopc_Ld5.T62_g16.GMOM.cheyenne_intel.mom-nuopc_cap       
        **expected - restart test passed - but mom history files were not in baseline**
```
**restart failures**
```
    FAIL ERS_Vmct_Ld5.T62_g16.GMOM.cheyenne_intel.mom-nuopc_cap                 
    FAIL ERR_Vmct_Ld5.f19_g16.BMOM.cheyenne_intel.allactive-nuopc_cap_io        
    FAIL ERS_Vmct_Ld5.T62_g16.CMOM.cheyenne_intel.mom-nuopc_cap                 
    FAIL ERS_Vmct_Ln9.f19_g16.X.cheyenne_intel                                  
       **expected based on previous failures**
    FAIL ERS_Vnuopc_Ld5.T62_g16.CMOM.cheyenne_intel.mom-nuopc_cap               
    FAIL ERS_Vnuopc_Ld5.T62_g16.GMOM.cheyenne_intel.mom-nuopc_cap               
        expected - restart test passed - but mom history files were not in baseline
```
Fixes #40 
User interface changes?: no
Update gh-pages html (Y/N)?:N

Code review: dunlap, jedwards

### CODE REVIEW

File: src/drivers/nuopc/shr/shr_nuopc_time_mod.F90 [hunk 1]
```diff
@@ -388,7 +385,7 @@ subroutine shr_nuopc_time_timeInit( Time, ymd, cal, tod, desc, logunit )
     ! local varaibles
     integer                     :: yr, mon, day ! Year, month, day as integers
     integer                     :: ltod         ! local tod
-    character(CL)               :: ldesc        ! local desc
+    character(len=256)          :: ldesc        ! local desc
```
Comment (jedwards4b) [inline src/drivers/nuopc/shr/shr_nuopc_time_mod.F90:388]: why this change?  Don't hard code string lengths.

**Review by jedwards4b — APPROVED** [review]
I'm overwhelmed, this is way too much for a single PR, but if Rocky can reproduce the test results then I'll accept it and move on.

**Comment by mvertens** [comment]
@jedwards4b - I agree. I will try to make my PRs smaller moving forwards. I
was waiting for you to get your PRs in and as a result this PR just kept
getting bigger.

**Comment by rsdunlapiv** [comment]
Tests are running:
```
dunlap@cheyenne1:~/UFSCOMP.aug27/cime/scripts> qcmd -l walltime=04:00:00 -- ./create_test --xml-testlist ./testlist_cmeps.xml --xml-category prealpha --xml-machine cheyenne --test-id aug27  --compare aug20 --baseline-root /glade/scratch_old/dunlap/BASELINES -o
Submitting command to PBS using account P93300606:
    ./create_test --xml-testlist ./testlist_cmeps.xml --xml-category prealpha --xml-machine cheyenne --test-id aug27 --compare aug20 --baseline-root /glade/scratch_old/dunlap/BASELINES -o
```

**Comment by mvertens** [comment]
Thank you!

**Comment by jedwards4b** [comment]
@rsdunlapiv if for some reason you need to restart those, please add the --generate flag

**Comment by rsdunlapiv** [comment]
Tests restarted:
```
dunlap@cheyenne3:~/UFSCOMP.aug27/cime/scripts> qcmd -l walltime=04:00:00 -- ./create_test --xml-testlist ./testlist_cmeps.xml --xml-category prealpha --xml-machine cheyenne --test-id aug28  --compare aug20 --generate aug28 --baseline-root /glade/scratch/dunlap/BASELINES -o
Submitting command to PBS using account P93300606:
    ./create_test --xml-testlist ./testlist_cmeps.xml --xml-category prealpha --xml-machine cheyenne --test-id aug28 --compare aug20 --generate aug28 --baseline-root /glade/scratch/dunlap/BASELINES -o
```

**Comment by jedwards4b** [comment]
Please update the original PR message with the correct cime hash @mvertens

**Comment by rsdunlapiv** [comment]
Failures were as expected:

```
dunlap@cheyenne6:/glade/scratch/dunlap> ./cs.status.aug28 | grep "FAIL "
    FAIL ERR_Vmct_Ld5.f19_g16.BMOM.cheyenne_intel.allactive-nuopc_cap_io NLCOMP
    FAIL ERR_Vmct_Ld5.f19_g16.BMOM.cheyenne_intel.allactive-nuopc_cap_io COMPARE_base_rest
    FAIL ERR_Vmct_Ld5.f19_g16.BMOM.cheyenne_intel.allactive-nuopc_cap_io BASELINE aug20
    FAIL ERR_Vnuopc_Ld5.f19_g16.BMOM.cheyenne_intel.allactive-nuopc_cap_io NLCOMP
    FAIL ERR_Vnuopc_Ld5.f19_g16.BMOM.cheyenne_intel.allactive-nuopc_cap_io COMPARE_base_rest
    FAIL ERR_Vnuopc_Ld5.f19_g16.BMOM.cheyenne_intel.allactive-nuopc_cap_io BASELINE aug20
    FAIL ERS_Vmct_Ld5.T62_g16.CMOM.cheyenne_intel.mom-nuopc_cap NLCOMP
    FAIL ERS_Vmct_Ld5.T62_g16.CMOM.cheyenne_intel.mom-nuopc_cap COMPARE_base_rest
    FAIL ERS_Vmct_Ld5.T62_g16.GMOM.cheyenne_intel.mom-nuopc_cap COMPARE_base_rest
    FAIL ERS_Vmct_Ln9.f19_g16.X.cheyenne_intel COMPARE_base_rest
    FAIL ERS_Vmct_Ln9.T31_g37_rx1.A.cheyenne_intel NLCOMP
    FAIL ERS_Vnuopc_Ld5.T62_g16.CMOM.cheyenne_intel.mom-nuopc_cap NLCOMP
    FAIL ERS_Vnuopc_Ld5.T62_g16.CMOM.cheyenne_intel.mom-nuopc_cap COMPARE_base_rest
    FAIL ERS_Vnuopc_Ld5.T62_g16.CMOM.cheyenne_intel.mom-nuopc_cap BASELINE aug20
    FAIL ERS_Vnuopc_Ld5.T62_g16.GMOM.cheyenne_intel.mom-nuopc_cap NLCOMP
    FAIL ERS_Vnuopc_Ld5.T62_g16.GMOM.cheyenne_intel.mom-nuopc_cap COMPARE_base_rest
    FAIL ERS_Vnuopc_Ld5.T62_g37.DTEST.cheyenne_intel.cice-nuopc_cap NLCOMP
    FAIL ERS_Vnuopc_Ln5.f19_g16.F2000Nuopc.cheyenne_intel.cam-nuopc_cap NLCOMP
    FAIL ERS_Vnuopc_Ln5.f45_f45_mg37.I2000Clm50SpNuopc.cheyenne_intel.clm-nuopc_cap NLCOMP
    FAIL ERS_Vnuopc_Ln5.ne16_ne16_mg17.QPC4.cheyenne_intel.cam-nuopc_cap NLCOMP
    FAIL ERS_Vnuopc_Ln9.f19_g16.X.cheyenne_intel NLCOMP
    FAIL ERS_Vnuopc_Ln9.f19_g16.X.cheyenne_intel BASELINE aug20
    FAIL ERS_Vnuopc_Ln9.T31_g37_rx1.A.cheyenne_intel NLCOMP
    FAIL SMS_Vmct_Ld1.T31_g37_rx1.A.cheyenne_intel NLCOMP
    FAIL SMS_Vmct_Ld5.f19_g16.BMOM.cheyenne_intel.allactive-nuopc_cap_io NLCOMP
    FAIL SMS_Vmct_Ld5.f19_g16.BMOM.cheyenne_intel.allactive-nuopc_cap_io BASELINE aug20
    FAIL SMS_Vmct_Ld5.T62_g16.CMOM.cheyenne_intel.mom-nuopc_cap NLCOMP
    FAIL SMS_Vnuopc.f19_g16.X.cheyenne_intel NLCOMP
    FAIL SMS_Vnuopc.f19_g16.X.cheyenne_intel BASELINE aug20
    FAIL SMS_Vnuopc_Ld1.T31_g37_rx1.A.cheyenne_intel NLCOMP
    FAIL SMS_Vnuopc_Ld1.T31_g37_rx1.A.cheyenne_intel BASELINE aug20
    FAIL SMS_Vnuopc_Ld5.f19_g16.BMOM.cheyenne_intel.allactive-nuopc_cap_io NLCOMP
    FAIL SMS_Vnuopc_Ld5.f19_g16.BMOM.cheyenne_intel.allactive-nuopc_cap_io BASELINE aug20
    FAIL SMS_Vnuopc_Ld5.T62_g16.CMOM.cheyenne_intel.mom-nuopc_cap NLCOMP
    FAIL SMS_Vnuopc_Ld5.T62_g16.CMOM.cheyenne_intel.mom-nuopc_cap BASELINE aug20
    FAIL SMS_Vnuopc_Ld5.T62_g16.GMOM.cheyenne_intel.mom-nuopc_cap NLCOMP
    FAIL SMS_Vnuopc_Ld5.T62_g37.DTEST.cheyenne_intel.cice-nuopc_cap NLCOMP
    FAIL SMS_Vnuopc_Ln5.f19_g16.F2000Nuopc.cheyenne_intel.cam-nuopc_cap NLCOMP
    FAIL SMS_Vnuopc_Ln5.f45_f45_mg37.I2000Clm50SpNuopc.cheyenne_intel.clm-nuopc_cap NLCOMP
    FAIL SMS_Vnuopc_Ln5.ne16_ne16_mg17.QPC4.cheyenne_intel.cam-nuopc_cap NLCOMP
```

**Pull request merged**